src/erc721.rs

```diff
@@ -0,0 +1,367 @@
+        };
+
+        return approved_as_owner || approved_as_user || approved_as_delegate;
+    }
+
+    // Mint a new token
+    pub(crate) fn _mint(to: &T::AccountId, token_id: &T::Hash) -> DispatchResult {
+        ensure!(!Self::_exists(token_id), "Token already exists");
+
+        let balance_of = Self::balance_of(to);
+
+        let new_balance_of = match balance_of.checked_add(1) {
+            Some(c) => c,
+            None => return Err(Error::<T>::Overflow.into()),
+        };
+
+        Self::_add_token_to_all_tokens_enumeration(token_id)?;
+        Self::_add_token_to_owner_enumeration(to, token_id)?;
+
+        <TokenOwner<T>>::insert(token_id, &to);
+        <OwnedTokensCount<T>>::insert(&to, new_balance_of);
+
+        Self::deposit_event(RawEvent::Transfer(None, Some(to.clone()), *token_id));
+
+        Ok(())
+    }
+
+    // Burn a token
+    pub(crate) fn _burn(token_id: &T::Hash) -> DispatchResult {
+        let owner = match Self::owner_of(token_id) {
+            Some(c) => c,
+            // None => return Err("No owner for this token"),
+            None => return Err(Error::<T>::TokenNoOwner.into()),
+        };
+
+        let balance_of = Self::balance_of(&owner);
+
+        let new_balance_of = match balance_of.checked_sub(1) {
+            Some(c) => c,
+            // None => return Err("Underflow subtracting a token to account balance"),
+            None => return Err(Error::<T>::Underflow.into()),
+        };
+
+        // Writing to storage begins here
+        Self::_remove_token_from_all_tokens_enumeration(token_id)?;
+        Self::_remove_token_from_owner_enumeration(&owner, token_id)?;
+        <OwnedTokensIndex<T>>::remove(token_id);
+
+        Self::_clear_approval(token_id)?;
+
+        <OwnedTokensCount<T>>::insert(&owner, new_balance_of);
+        <TokenOwner<T>>::remove(token_id);
+
+        Self::deposit_event(RawEvent::Transfer(Some(owner), None, *token_id));
+
+        Ok(())
+    }
+
+    // Transfer token from one account to other
+    pub(crate) fn _transfer_from(
+        from: &T::AccountId,
+        to: &T::AccountId,
+        token_id: &T::Hash,
+    ) -> DispatchResult {
+        let owner = match Self::owner_of(token_id) {
+            Some(c) => c,
+            None => return Err(Error::<T>::TokenNoOwner.into()),
+        };
+
+        ensure!(owner == *from, "'from' account does not own this token");
+
+        let balance_of_from = Self::balance_of(from);
+        let balance_of_to = Self::balance_of(to);
+
+        let new_balance_of_from = match balance_of_from.checked_sub(1) {
+            Some(c) => c,
+            None => return Err(Error::<T>::Underflow.into()),
+        };
+
+        let new_balance_of_to = match balance_of_to.checked_add(1) {
+            Some(c) => c,
+            None => return Err(Error::<T>::Overflow.into()),
+        };
+
+        Self::_remove_token_from_owner_enumeration(from, token_id)?;
+        Self::_add_token_to_owner_enumeration(to, token_id)?;
```

Comment:
1 the potential error is from number overflow or underflow. Actually, overflow almost impossible since we give a u64 integer. And underflow could be avoid if we always put a increment before a decrement. I have noticed some pallet neglect such error. Just do increment/decrement operations.
2. I think the real state change will be applied after the whole extrinsic successfully executed in WASM EM. I am not sure it will get a broken state. I will try to verify in real runtime.

Reply:
See above